FEAT: Validate that the output type of a UDF is a single element

timothydijamco opened the pull request:
The normal syntax for defining an elementwise UDF is passing a single value to `output_type`:
```
@elementwise(input_type=[dt.float64, dt.float64], output_type=dt.float64)
def sum_udf(a, b):
    return a + b
```


If the output type is placed into a list (similar to `input_type`) (see below), then executing an expression with the UDF succeeds on Pandas but fails on Spark with `TypeError: 'NoneType' object is not callable`
```
@elementwise(input_type=[dt.float64, dt.float64], output_type=[dt.float64])
def sum_udf(a, b):
    return a + b
```


Currently, `output_type=[dt.float64]` is interpreted as the output type being `array<float64>`

This PR unpacks the single value in `output_type` if it's a list and uses that as the output type so that e.g. `output_type=[dt.float64]` is treated as just a `float64`



**edit:** This PR was originally "BUG: Allow output type of vectorized UDFs to be specified as a single-element list" but was renamed (see below)

timothydijamco commented:
cc @icexelloss @jreback

icexelloss commented on ibis/tests/all/test_vectorized_udf.py:
```diff
@@ -10,9 +10,22 @@ def add_one(s):
 
 
+@elementwise(input_type=[dt.double], output_type=[dt.double])
+def add_one_with_output_type_list(s):
```
I think you can inline this function in the tests, we don't think we need to use this in other tests.

icexelloss commented on ibis/tests/all/test_vectorized_udf.py:
```diff
@@ -10,9 +10,22 @@ def add_one(s):
+
+@pytest.mark.only_on_backends([Pandas, PySpark])
+@pytest.mark.xfail_unsupported
+def test_elementwise_udf_with_output_type_list(backend, alltypes, df):
```
We should probably comment that this does NOT test the output of the UDF being and list. We only test that specifying output=[dt] is the same as output=dt in the udf decorator.

icexelloss commented on ibis/udf/vectorized.py:
```diff
@@ -18,8 +18,13 @@ class UserDefinedFunction(object):
+
+        self.input_type = list(map(dt.dtype, input_type))
+
+        if type(output_type) is list:
```
It's more common to use `isinstance(output_type, list)`

icexelloss commented on ibis/udf/vectorized.py:
```diff
@@ -18,8 +18,13 @@ class UserDefinedFunction(object):
+        self.input_type = list(map(dt.dtype, input_type))
+
+        if type(output_type) is list:
+            [output_type] = output_type
```
It's more comment to use 

`output_type, = output_type`

icexelloss replied:
should probably also catch the exception when the list has more then one element and give a reasonable error message

timothydijamco replied:
> It's more comment to use
> 
> `output_type, = output_type`

blacken was changing this to `(output_type,) = output_type` so `[output_type] = output_type` seemed cleaner, however, if `output_type,` is more common then maybe `(output_type,)` is more readable


> should probably also catch the exception when the list has more then one element and give a reasonable error message

Will do

jreback commented:
if you can rebase this

timothydijamco commented:
Done! CI is green

The rebase did involve resolving some merge conflicts, but the changes are exactly the same besides the name of the test and the format of the test's comment. Because of #2210 the changed now apply also to `analytic` and `reduction` UDFs, so I'll rename the PR

jreback commented:
hmm, why are we allowing this kind of output type as a list? I would rather just raise if this is specified (on all backends) as better to be strict.

icexelloss commented:
@jreback I am fine with raise. But I prefer allowing `output_type=[dt.double]` a little more because this is more symmetrical to the input_type parameter. e.g.

```
@elementwise(
    input_type=[dt.double, dt.double]
    output_type=[dt.double]
)
def foo(...)
```

I know currently output type can only be a single col, but it seems natural to type `output_type=[dt.double]` than `output_type=dt.double` after typing `input_type=[...]`

timothydijamco commented:
^ Sounds good to me!

I refactored the validation function in [`ibis/udf/vectorized.py`](https://github.com/ibis-project/ibis/blob/d24f0d88f36731a50f555956cf69bea4e9b037db/ibis/udf/vectorized.py) out into a new module, [`ibis/udf/validate.py`](https://github.com/ibis-project/ibis/blob/d24f0d88f36731a50f555956cf69bea4e9b037db/ibis/udf/validate.py) and added a function to that module to validate the `output_type`. That function is now used throughout the backends that have UDFs with output type arguments

This is starting to sound more like a feature (adding validation) than a bug fix to me, so I'm going to rename this PR

icexelloss commented on ibis/udf/vectorized.py:
```diff
@@ -102,6 +52,10 @@ def func(*args):
 
 
 def _udf_decorator(node_type, input_type, output_type):
+    if isinstance(output_type, list):
```
Please update https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L3462 to do proper rlz checking instead of checking it here.

See https://github.com/ibis-project/ibis/blob/master/ibis/expr/rules.py and other examples in https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py

timothydijamco replied:
Oops, this shouldn't be here (the checking is already done in UserDefinedFunction `__init__`)

timothydijamco replied:
I'll look into doing checking using rlz

timothydijamco replied:
I was trying rlz checking, but now I think it actually might be more appropriate to do the checking before we create the node (in other words, not using rlz)

Using rlz seems to be geared towards describing/validating each argument of a node, when the argument is in its "final form" in the node

**E.g.**
We want to make sure `output_type=[dt.double]` is caught and an error is raised. (Note that `output_type` "starts out" as a Python `list`)

So we can detect this syntax by (at some point) checking if the `output_type` is a Python `list`

But by the time we're in the node doing rlz checking, `output_type` would have been converted to an Ibis datatype [here](https://github.com/ibis-project/ibis/blob/master/ibis/udf/vectorized.py#L106). The rule would see `output_type` as an Ibis `Array`, which I think is valid (user could have intentionally specified `output_type=dt.Array(dt.double)`) so we can't raise an error

timothydijamco replied:
There are workarounds that would let us still use a rlz check (e.g. do the type conversion in the node after validation has been done) but after exploring this I feel like it makes sense to conceptually treat the validation as checking the user's input to the UDF API as opposed to validating the node arguments

timothydijamco replied:
(I still plan to remove this particular check in `_udf_decorator`, it's redundant (the checking is already done [elsewhere](https://github.com/ibis-project/ibis/blob/510f68d91c90d76e8958107765e383ca6088b46b/ibis/udf/vectorized.py#L28)))

timothydijamco replied:
Done!

icexelloss commented on ibis/udf/validate.py:
```diff
@@ -0,0 +1,71 @@
+import ibis.common.exceptions as com
+
+
+def _parameter_count(funcsig):
```
Why take this out? The origin file is not very large.

timothydijamco replied:
I think this depends on whether we want to do this for vectorized UDFs only or all backends. I moved this out so that the import wouldn't be `ibis.udf.vectorized` the functions in here

timothydijamco replied:
(see https://github.com/ibis-project/ibis/pull/2198#discussion_r432728880)

icexelloss commented on ibis/tests/all/test_vectorized_udf.py:
```diff
@@ -5,14 +5,13 @@
 @pytest.mark.only_on_backends([Pandas, PySpark])
 @pytest.mark.xfail_unsupported
 def test_elementwise_udf(backend, alltypes, df):
+    @elementwise(input_type=[dt.double], output_type=dt.double)
```
Why move inside?

timothydijamco replied:
Oops, I had moved this in when I created the test that checks that `output_type` works the same (I've removed that test now though). I can revert this

icexelloss replied:
Yeah let's revert unnecessary changes

timothydijamco replied:
I added a test to check that there's an error when `output_type` is a list and it has an inline `add_one` UDF, so I'm keeping this change (moving global `add_one` to be inline inside `test_elementwise_udf`)

icexelloss commented on ibis/sql/postgres/udf/api.py:
```diff
@@ -90,6 +92,8 @@ def existing_udf(name, input_types, output_type, schema=None, parameters=None):
             ).format(len(input_types), len(parameters))
         )
 
+    v.validate_output_type(output_type)
```
postgres udf doesn't necessarily need to follow the restriction of `ibis.udf.vectorized` we should not change anything about postgres udf in this PR

timothydijamco replied:
Postgres the some other backends also have UDFs with the pattern of having an `input_type` list and single `output_type` so these changes were to do validation for those too (basically, I was expanding the PR to apply to all UDFs in general with this change not just vectorized UDFs)

icexelloss replied:
Let's limit the scope to Pandas and Pyspark for now.

jreback replied:
i am ok with validating everything, that in fact is the premise that we have the same udf signatures across backends, which AFAICT is true.

timothydijamco replied:
> Let's limit the scope to Pandas and Pyspark for now.

This made sense to me (adding validation only to vectorized UDFs) especially since if the validation should be done in the node using rlz, it easily can't be applied to UDFs in other backends

But I've started to think it is better to have the validation outside of the node (basically, where the validation checks are now in this PR) (see https://github.com/ibis-project/ibis/pull/2198#discussion_r433911429)

In that case, UDFs in other backends can be validated pretty easily and using the same function as we use for vectorized UDFs, `validate_output_type` (the current state of this PR does this already, so validating everything would just mean leaving the changes in the other backends as-is instead of reverting them)

I feel like it makes sense to leave in the validation for all UDFs in this PR. what do you think?

timothydijamco replied:
> i am ok with validating everything, that in fact is the premise that we have the same udf signatures across backends, which AFAICT is true.

From what I saw most of the backends have UDFs that have signatures that are pretty similar (but not exactly the same) (bigquery, impala, and spark have `input_type` and `output_type`). [SQLite's UDF](https://github.com/ibis-project/ibis/blob/master/ibis/sql/sqlite/client.py#L28) doesn't have `output_type` though

icexelloss replied:
Ok with validating.

icexelloss commented on ibis/spark/udf.py:
```diff
@@ -32,9 +32,10 @@ def validate_func_and_types(self, func):
-        valid_function_signature(self.input_type, func)
+        # Validate that the input_type argument and the function signature
+        # match and that the output_type is valid
+        v.validate_input_type_count(self.input_type, func)
```
Don't need to touch Spark udf either. Only PySpark and Pandas backend supports "vectorized UDF".

icexelloss commented on ibis/bigquery/udf/api.py:
```diff
@@ -159,6 +160,8 @@ class Rectangle {
     return my_rectangle(width, height);
     """;
     '''
+    v.validate_output_type(output_type)
```
No need to change big query

icexelloss replied:
Discussed offline. OK to add validation here.

jreback commented:
@timothydijamco can you also rebase

icexelloss commented on ibis/udf/validate.py:
```diff
@@ -0,0 +1,71 @@
+    )
+
+
+def validate_input_type_count(input_type, func):
```
I think `validate_input_type` is better here

timothydijamco commented:
@jreback @icexelloss 

Rebased and addressed comments

and CI is green!

jreback commented on ibis/udf/validate.py:
```diff
@@ -0,0 +1,71 @@
+import ibis.common.exceptions as com
+
+
+def _parameter_count(funcsig):
```
can you type this

jreback commented on ibis/udf/validate.py:
```diff
@@ -0,0 +1,71 @@
+    )
+
+
+def validate_input_type(input_type, func):
```
can you type this

jreback commented on ibis/udf/validate.py:
```diff
@@ -0,0 +1,71 @@
+    return funcsig
+
+
+def validate_output_type(output_type):
```
can you type this

pep8speaks commented:
Hello @timothydijamco! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`ibis/sql/postgres/udf/api.py`](https://github.com/ibis-project/ibis/blob/08d17440dbd4fba73cf54069a4cf433c0a4c15a2/ibis/sql/postgres/udf/api.py):

> [Line 95:80](https://github.com/ibis-project/ibis/blob/08d17440dbd4fba73cf54069a4cf433c0a4c15a2/ibis/sql/postgres/udf/api.py#L95): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)

timothydijamco commented:
@jreback CI is green

Also I'm not sure what the PEP warning is above--the line that it links to is short

jreback commented:
thanks @timothydijamco 

so the PEP warning above is from the configured pep8speaks, but i don't think it uses the same settings as the project. can you open an issue about this? (I think we just need to add a .pep8speaks) or something

timothydijamco commented:
Thank you both
